### ISSUE TITLE
feat: remove `readonly` constraint from response related types

### ISSUE THREAD

**Issue by nunomaduro** [algolia/algoliasearch-client-javascript#1064]
**TL; DR**: Go to every type that ends with `...Response.ts` and remove `readonly` keywords.

A good type-system prevents problems from happening - that's why, by default, the given params to a async function in the client most be *readonly*. So they can't be modified while the promise is being resolved in background. Here is an example:

```ts
client.search('foo', params).then(console.log);
params.foo = 'bar'; // TYPE-ERROR: because those params may still be used inside the `search` function.
```

In the other hand, having `readonly` responses prevents the user from doing manipulations to the response object, and that's is annoying:

```ts
const response await client.search('foo', params);
delete response.hits; // TYPE-ERROR: grrrrrr
const myCustomObject = response;
```

Since having `readonly` responses add no value to the user or to us, to fix this problem, go to every type that ends with `...Response.ts` and remove `readonly` keywords.

**Comment by nunomaduro** [comment]
cc @francoischalifour

**Comment by nunomaduro** [comment]
@francoischalifour Maybe would be cool to have actually fixing the problem. Is a good first issue for having you in touch with the source code.

**Comment by Haroenv** [comment]
Could we maybe expose it behind a generic? i.e. if you pass some parameters (not sure what), we could do `type DeepWriteable<T> = { -readonly [P in keyof T]: DeepWriteable<T[P]> };` on the result type

Or the other way round, `search<Mutability = ReadOnly> { return Mutability<xxx> }` where you could pass the `DeepWriteable` manually?

Not sure if either of those options are something users think about though

**Comment by nunomaduro** [comment]
@Haroenv You will have to explain me that in a zoom call, sorry.
